File: packages/oxlint-plugin-react-doctor/src/plugin/utils/is-nullish-expression.ts
import type { EsTreeNode } from "./es-tree-node.js";
import { isNodeOfType } from "./is-node-of-type.js";

// True for the statically-nullish expression shapes: the `null` literal,
// the bare `undefined` identifier, and a `void …` UnaryExpression (which
// always evaluates to `undefined`). React renders all three as nothing,
// and none can carry a prop value.
export const isNullishExpression = (expression: EsTreeNode): boolean =>
  (isNodeOfType(expression, "Literal") && expression.value === null) ||
  (isNodeOfType(expression, "Identifier") && expression.name === "undefined") ||
  (isNodeOfType(expression, "UnaryExpression") && expression.operator === "void");

File: native/oxlint/rules/is-nullish-expression.rs
fn is_nullish_expression(expression: &oxc_ast::ast::Expression<'_>) -> bool {
    match expression.get_inner_expression() {
        oxc_ast::ast::Expression::NullLiteral(_) => true,
        oxc_ast::ast::Expression::Identifier(identifier) => identifier.name == "undefined",
        oxc_ast::ast::Expression::UnaryExpression(unary_expression) => {
            unary_expression.operator == oxc_syntax::operator::UnaryOperator::Void
        }
        _ => false,
    }
}
